src/main/java/com/iabtcf/FieldDefs.java

```diff
@@ -0,0 +1,427 @@
+     */
+    private static abstract class MemoizingFunction
+            implements LengthSupplier, OffsetSupplier, Function<ByteBitVector, Integer> {
+        private boolean dynamicInitialized = false;
+        private boolean isDynamic = false;
+        private Integer value;
+
+        public abstract Integer doCompute(ByteBitVector t);
+
+        @Override
+        public abstract boolean isDynamic();
+
+        @Override
+        public Integer apply(ByteBitVector t) {
+            if (value == null || isDynamicPvt()) {
+                value = doCompute(t);
+
+                return value;
+            } else {
+                return value;
+            }
+        }
+
+        private boolean isDynamicPvt() {
+            if (!dynamicInitialized) {
+                isDynamic = isDynamic();
+            }
+
+            return isDynamic;
+        }
+    }
+
+    private interface OffsetSupplier extends Function<ByteBitVector, Integer> {
+
+        /**
+         * This is used when we don't want a field to support offsets.
+         */
+        public static final OffsetSupplier NOT_SUPPORTED = new OffsetSupplier() {
+
+            @Override
+            public Integer apply(ByteBitVector t) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public boolean isDynamic() {
+                return false;
+            }
+        };
+
+        /**
+         * A constant offset for static fields.
+         */
+        public static OffsetSupplier constant(int offset) {
+            return new OffsetSupplier() {
+
+                @Override
+                public Integer apply(ByteBitVector t) {
+                    return offset;
+                }
+
+                @Override
+                public boolean isDynamic() {
+                    return false;
+                }
+            };
+        }
+
+        /**
+         * Supplies the offset that's based on the specified field. The offset value is stored to
+         * avoid re-computing.
+         */
+        public static OffsetSupplier from(final FieldDefs thisEnum) {
+            return new MemoizingFunction() {
+                @Override
+                public boolean isDynamic() {
+                    return thisEnum.isDynamic();
+                }
+
+                @Override
+                public Integer doCompute(ByteBitVector t) {
+                    return thisEnum.getLength(t) + thisEnum.getOffset(t);
+                }
+            };
+        }
+
+        /**
+         * Supplies the offset that's based on the the fields previous field. The offset value is
+         * stored to avoid re-computing.
+         */
+        public static OffsetSupplier fromPrevious(final FieldDefs thisEnum) {
+            return new MemoizingFunction() {
+
+                @Override
+                public boolean isDynamic() {
+                    return FieldDefs.values()[thisEnum.ordinal() - 1].isDynamic();
```

Comment:
This goes back to my other point of using ordinal. This makes this code VERY brittle. Especially since you have to take a deep dive into this code to realize that if you only pass a length to the enum constructor it will give you an OffsetSupplier that reads the fields in the order they were defined in this enum, but only up to a point that the next previous field did not use pass an offset supplier to its constructor. I would much rather see the fields that use this logic to have a statically defined offset the is visible and readable. That would reduce the code in this class and can still be checked by unit tests in the same way that this is now.

Reply:
this is by design